File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    answer = []
    
    # 사전 초기화
    dt = {}
    for i in range(ord('A'), ord('Z') + 1):
        dt[chr(i)] = i - ord('A') + 1
        
    idx_num = 27  # 색인 번호
    for i in range(0, len(msg)):
        if i == 0:
            curr = msg[0]
            continue
            
        next_msg = curr + msg[i]
        if next_msg in dt:
            curr = next_msg
            
        else:
            answer.append(dt[curr])
            dt[next_msg] = idx_num
            idx_num += 1
            curr = msg[i]

    # 마지막으로 처리되지 않은 글자 처리
    answer.append(dt[curr])

    return answer

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.js
function solution(msg) {
    const map = new Map()
    
    // 색인 번호 생성
    for (let i = ('A').charCodeAt(); i <= ('Z').charCodeAt(); i++) {
        map.set(String.fromCharCode(i), i - ('A').charCodeAt() + 1)
    }
    
    const answer = []
    let curr_idx = 27
    let curr_msg
    for (let i = 0; i < msg.length; i++) {
        if (i === 0) {
            curr_msg = msg[0]
            continue
        }
        
        let next_msg = curr_msg + msg[i]
        if (!map.get(next_msg)) {
            map.set(next_msg, curr_idx++)
            answer.push(map.get(curr_msg))
            curr_msg = msg[i]
        } else {
            curr_msg = next_msg
        }
    }
    
    // 마지막으로 처리되지 않은 문자 처리
    answer.push(map.get(curr_msg))
    
    return answer
}